src/Features/Core/Portable/CodeRefactorings/ConvertConversionsOperators/AbstractConvertConversionOperatorsRefactoringProvider.cs

```diff
@@ -0,0 +1,105 @@
+            var asExpressions = await context.GetRelevantNodesAsync<TAsExpressionSyntax>().ConfigureAwait(false);
+
+            if (castExpressions.IsEmpty && asExpressions.IsEmpty)
+            {
+                return;
+            }
+
+            var (document, cancellationToken) = (context.Document, context.CancellationToken);
+            var semanticModelFactory = AsyncLazy.Create(async cancellationToken
+                => await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false), cacheResult: true);
+
+            castExpressions = await FilterCastExpressionCandidatesAsync(castExpressions, semanticModelFactory, cancellationToken).ConfigureAwait(false);
+            asExpressions = await FilterAsExpressionCandidatesAsync(asExpressions, semanticModelFactory, cancellationToken).ConfigureAwait(false);
+
+            if (castExpressions.IsEmpty && asExpressions.IsEmpty)
+            {
+                return;
+            }
+
+            var nodesToRegister = castExpressions.Cast<SyntaxNode>().Union(asExpressions).ToImmutableArray();
+
+            foreach (var node in nodesToRegister)
+            {
+                Func<CancellationToken, Task<Document>> createChangedDocument = node switch
+                {
+                    TCastExpressionSyntax castExpression => cancellationToken => ConvertFromCastToAsAsync(document, castExpression, cancellationToken),
+                    TAsExpressionSyntax asExpression => cancellationToken => ConvertFromAsToCastAsync(document, asExpression, cancellationToken),
+                    _ => throw new InvalidOperationException("Unsupported node type"),
+                };
+                context.RegisterRefactoring(
+                    new MyCodeAction(
+                        GetTitle(),
+                        createChangedDocument
+                    ), node.Span);
+            }
+        }
+
+        protected virtual Task<ImmutableArray<TCastExpressionSyntax>> FilterCastExpressionCandidatesAsync(
+            ImmutableArray<TCastExpressionSyntax> castExpressions,
+            AsyncLazy<SemanticModel> semanticModelFactory,
+            CancellationToken cancellationToken)
+            => Task.FromResult(castExpressions);
+
+        protected virtual Task<ImmutableArray<TAsExpressionSyntax>> FilterAsExpressionCandidatesAsync(
+            ImmutableArray<TAsExpressionSyntax> asExpressions,
+            AsyncLazy<SemanticModel> semanticModelFactory,
+            CancellationToken cancellationToken)
+            => Task.FromResult(asExpressions);
+
+        protected abstract Task<Document> ConvertFromCastToAsAsync(
+            Document document,
+            TCastExpressionSyntax castExpression,
+            CancellationToken cancellationToken);
+
+        protected abstract Task<Document> ConvertFromAsToCastAsync(
+            Document document,
+            TAsExpressionSyntax asExpression,
+            CancellationToken cancellationToken);
```

Comment:
i see no benefit in having one refactoring provider that has to handle both node types and hten switch and handle each separately.  just have two different code refactoring providers.  one from Cast->As and one from As->Cast.